src/HealthGPS/energy_balance_model.cpp

```diff
@@ -78,131 +82,169 @@ void EnergyBalanceModel::update_risk_factors(RuntimeContext &context) {
+                              .F = F,
+                              .L = L,
+                              .ECF = ECF,
+                              .G = G,
+                              .W = W,
+                              .EE = EE,
+                              .EI = EI,
+                              .adjust = adjust};
+
+    return state;
+}
+
+std::unordered_map<core::Identifier, double>
+EnergyBalanceModel::compute_nutrients(const Person &person) const {
+    std::unordered_map<core::Identifier, double> nutrient_intakes;
+
+    for (const auto &equation : nutrient_equations_) {
+        double food_intake = person.get_risk_factor_value(equation.first);
+
+        for (const auto &coefficient : equation.second) {
+            double delta_nutrient = food_intake * coefficient.second;
+            nutrient_intakes[coefficient.first] += delta_nutrient;
+        }
+    }
+
+    return nutrient_intakes;
+}
+
+double EnergyBalanceModel::compute_EI(
+    const std::unordered_map<core::Identifier, double> &nutrient_intakes) const {
+    double EI = 0.0;
 
-    // Update energy expenditure.
-    double EE =
-        (final_shift + K + gamma_F * F + gamma_L * L + delta_BW + TEF + AT + EI * x) / (1.0 + x);
-
-    if (final_run) { // Final run:
-        // Save state and return nothing.
-        person.risk_factors[H_key] = H;
-        person.risk_factors[BW_key] = BW;
-        person.risk_factors[PAL_key] = PAL;
-        person.risk_factors[F_key] = F;
-        person.risk_factors[L_key] = L;
-        person.risk_factors[ECF_key] = ECF;
-        person.risk_factors[EI_key] = EI;
-        person.risk_factors[EE_key] = EE;
-        return std::array<double, 2>{};
-    } else { // Trial run:
-        // return BW and baseline adjustment coefficient.
-        double trial_adjust = -(a1 - b1) * (1.0 - exp(-365.0 / tau)) / (a1 * b2 - a2 * b1);
-        return std::array<double, 2>{BW, trial_adjust};
+    for (const auto &coefficient : energy_equation_) {
+        double delta_energy = nutrient_intakes.at(coefficient.first) * coefficient.second;
+        EI += delta_energy;
     }
+
+    return EI;
+}
+
+double EnergyBalanceModel::compute_G(double CI, double CI_0, double G_0) const {
+    double k_G = CI_0 / (G_0 * G_0);
+    double G = sqrt(CI / k_G);
+    return G;
+}
+
+double EnergyBalanceModel::compute_ECF(double EI, double EI_0, double CI, double CI_0,
+                                       double ECF_0) const {
+    double Na_b = 4000.0;
+    double Na_f = Na_b * EI / EI_0;
+    double Delta_Na_diet = Na_f - Na_b;
+    double ECF = ECF_0 + (Delta_Na_diet - xi_CI * (1.0 - CI / CI_0)) / xi_Na;
+    return ECF;
+}
+
+double EnergyBalanceModel::compute_RMR(double BW, double H, unsigned int age,
+                                       core::Gender gender) const {
+    double RMR = 9.99 * BW + 6.25 * H * 100.0 - 4.92 * age;
+    RMR += gender == core::Gender::male ? 5.0 : -161.0;
+    RMR *= 4.184; // kcal to kJ
+    return RMR;
```

Comment:
In case you're interested, I've used this library before for unit conversions in C++: https://github.com/nholthaus/units

It gives you compile-time dimensional analysis, which is quite neat. Maybe worth considering?

Reply:
I had this in mind, actually. I just want Ali to go over it and confirm all the units first, though. Some of them aren't so obvious.